### ISSUE TITLE
Linux: frequent crashes on startup

### ISSUE THREAD

**Issue by RfzVd** [blushiemagic/tModLoader#237]
### Description

The Re-Logic intro screen comes up, fades away, and the screen stays black after. The process has to be killed manually.

### Steps to Reproduce

1. Start the game from Steam
2. Hope for the best

### Expected behavior 

The tModLoader "finding mods"  screen should come up.

### Actual behavior

Doesn't.

### Reproduction frequency

~60-80% of startups.

### Versions

This happens on 0.10.0.2, don't know if happens on other versions.

### Mods

Happens without mods enabled.

### Additional Information

I tested without tModLoader about 6 times and the game starts fine always. With tModLoader it only works seldom. I can't find any error logs for the crashes/freezes.

I've had 2 ingame crashes and one of those produced a log in ~/.steam/steam/steamapps/common/Terraria/client-crashlog.txt. I use the Calamity mod and I don't know if it's related to it, tModLoader or something else but I'll include the log here. I'm not sure this warrants another issue.

8/20/2017 8:24:41 AM
System.TypeLoadException: Failure has occurred while loading a type.
  at Terraria.Social.Steam.CoreSocialModule.Initialize () [0x00039] in <6bf9c39ff4674fd3a65b4dd3174fc083>:0 
  at Terraria.Social.SocialAPI.Initialize (System.Nullable`1[T] mode) [0x00053] in <6bf9c39ff4674fd3a65b4dd3174fc083>:0 
  at Terraria.Program.LaunchGame (System.String[] args, System.Boolean monoArgs) [0x000d4] in <6bf9c39ff4674fd3a65b4dd3174fc083>:0

**Comment by Lukaboot** [comment]
No mention of your distro ? that's kind of a wild guess for them then

**Comment by RfzVd** [comment]
Oh, distro is Bunsenlabs, that is basically Debian Jessie with openbox. If there's any more possibly relevant information I can of course provide it.

**Comment by PoroCYon** [comment]
@RfzVd Check whether `Steamworks.NET.dll`, `libCSteamworks.so` and `libsteam_api.so` *all* exist. If not, validate.

**Comment by rolzor** [comment]
This also happens for me on mac OS 10.11.6 - running tmodloader 0.10.1

I've never left it running when the black screen happens, I usually just force quit and try again via steam. Sometimes it takes 4-5 tries before it actually loads.

My system.log file has lines which look like the following (I'm unsure if they're relevant or not):

   WindowServer[175]: disable_update_timeout: UI updates were forcibly disabled by application "Terraria" for over 1.00 seconds. Server has re-enabled them.
   WindowServer[175]: common_reenable_update: UI updates were finally reenabled by application "Terraria" after 7.98 seconds [0.13fps] (server forcibly re-enabled them after 1.00 seconds [1.00fps])

**Comment by MrLavender** [comment]
Same issue on mac OS 10.10.5 with tmodloader 0.10.1.1

It seems to be caused by starting up in fullscreen mode.

Editing ~/Library/Application Support/Terraria/ModLoader/config.json and changing the "Fullscreen" parameter to false resolves the issue for me. After the game has finished loading and is at the main menu then I can switch to fullscreen without any problems. Just have to remember to switch back to windowed mode before quitting...

**Comment by muesli4** [comment]
Can confirm with Linux version (native steam) on Archlinux. For me it works all the time after a restart. Maybe something related to graphic driver? (I'm using nvidia.)

@PoroCYon I'm missing `Steamworks.NET.dll`, but have `libCSteamworks.so` and `libsteam_api.so`. Repairing on steam will clear tModLoader (1 file replaced).

**Comment by PoroCYon** [comment]
muesli4: `Steamworks.NET.dll` is embedded inside `Terraria.exe` and must be manually extracted. The other dynamic libs you listed run *native code*, while there must be a wrapper present (which, in this case, should be `Steamworks.NET.dll`).

Also, try getting a stack trace, somehow (see also the `mono(1)` manpage). I doubt it is a driver issue, but you never know with nvidia `¯\_(ツ)_/¯`. If your GPU is supported by nouveau, you should try that instead.

Here's how to extract `Steamworks.NET.dll`:

```sh
$ csharp # execute this in the directory `Terraria.exe` resides in
> using System.IO;
> using System.Reflection;
> var asm = Assembly.LoadFrom("Terraria.exe");
> var str = asm.GetManifestResourceStream("Terraria.Libraries.Steamworks.NET.Linux.Steamworks.NET.dll");
> var fs = File.OpenWrite("Steamworks.NET.dll");
> str.CopyTo(fs);
> fs.Close();
> quit;
```

**Comment by StaticDynamics** [comment]
I had this problem earlier today and was able to get it to work by having Terraria disable compositing on startup using plasma's window settings or changing the compositing mode to xrender. It also worked fine when running it using integrated intel graphics instead of amd. Sorry if this is a bit hard to follow and unhelpful, but I'm just putting my experience here.

**Comment by PoroCYon** [comment]
This *could* be a compositor issue, but I'm not entirely sure. I had to change some stuff in FNA and change some environment variables in order to get Terraria to work on my (rather exotic) setup. You could try these and report whether they worked. (If it doesn't, try tweaking the environment variables.)

Envvars (i.e. how I run Terraria):
```sh
# FNA_SONG_DISABLE_THREADS and FNA_KEYBOARD_USE_SCANCODES aren't crucial for this
#FNA_SONG_DISABLE_THREADS=1 FNA_KEYBOARD_USE_SCANCODES=1 \
LD_LIBRARY_PATH="$(pwd)/lib64:${LD_LIBRARY_PATH:-/usr/lib}" \
FNA_OPENGL_DISABLE_LATESWAPTEAR=1 FNA_OPENGL_FORCE_CORE_PROFILE=1 \
FNA_GRAPHICS_FORCE_GLDEVICE=ModernGLDevice \
mono --debug=mdb-optimizations Terraria.exe # --debug flag used for better stacktraces
```

Change `LD_LIBRARY_PATH` to the path where the native libraries reside, `FNA_OPENGL_DISABLE_LATESWAPTEAR` and maybe `FNA_OPENGL_FORCE_CORE_PROFILE` and `FNA_GRAPHICS_FORCE_GLDEVICE` should be tweaked if it doesn't work.

Here's the patch: (Apply this to the FNA source, then recompile.)
```patch
diff --git a/src/FNAPlatform/ModernGLDevice.cs b/src/FNAPlatform/ModernGLDevice.cs
index f89ab62..91df36a 100644
--- a/src/FNAPlatform/ModernGLDevice.cs
+++ b/src/FNAPlatform/ModernGLDevice.cs
@@ -8,7 +8,7 @@
 #endregion
 
 #region THREADED_GL Option
-// #define THREADED_GL
+#define THREADED_GL
 /* Ah, so I see you've run into some issues with threaded GL...
  *
  * This class is designed to handle rendering coming from multiple threads, but
@@ -24,7 +24,7 @@
 #endregion
 
 #region DISABLE_THREADING Option
-// #define DISABLE_THREADING
+//#define DISABLE_THREADING
 /* Perhaps you read the above option and thought to yourself:
  * "Wow, only an idiot would need threads for their graphics code!"
  *
diff --git a/src/FNAPlatform/OpenGLDevice.cs b/src/FNAPlatform/OpenGLDevice.cs
index 9ae0d7f..5c096a2 100644
--- a/src/FNAPlatform/OpenGLDevice.cs
+++ b/src/FNAPlatform/OpenGLDevice.cs
@@ -8,7 +8,7 @@
 #endregion
 
 #region THREADED_GL Option
-// #define THREADED_GL
+#define THREADED_GL
 /* Ah, so I see you've run into some issues with threaded GL...
  *
  * This class is designed to handle rendering coming from multiple threads, but
@@ -24,7 +24,7 @@
 #endregion
 
 #region DISABLE_THREADING Option
-// #define DISABLE_THREADING
+//#define DISABLE_THREADING
 /* Perhaps you read the above option and thought to yourself:
  * "Wow, only an idiot would need threads for their graphics code!"
  *
diff --git a/src/FNAPlatform/SDL2_FNAPlatform.cs b/src/FNAPlatform/SDL2_FNAPlatform.cs
index dfa7937..511ef92 100644
--- a/src/FNAPlatform/SDL2_FNAPlatform.cs
+++ b/src/FNAPlatform/SDL2_FNAPlatform.cs
@@ -868,11 +868,11 @@ namespace Microsoft.Xna.Framework
 				{
 					if (SDL.SDL_GL_SetSwapInterval(-1) != -1)
 					{
-						FNALoggerEXT.LogInfo("Using EXT_swap_control_tear VSync!");
+						//FNALoggerEXT.LogInfo("Using EXT_swap_control_tear VSync!");
 					}
 					else
 					{
-						FNALoggerEXT.LogInfo("EXT_swap_control_tear unsupported. Fall back to standard VSync.");
+						//FNALoggerEXT.LogInfo("EXT_swap_control_tear unsupported. Fall back to standard VSync.");
 						SDL.SDL_ClearError();
 						SDL.SDL_GL_SetSwapInterval(1);
 					}
```

**Comment by Chicken-Bones** [comment]
Should be fixed https://github.com/blushiemagic/tModLoader/commit/09518af7ede1711bcee902a1fe5ae9c7bbd62210

**Comment by AvianGang** [comment]
Still is broken for me... not fixed?

**Comment by JavidPack** [comment]
It's fixed in the next release.

**Comment by jvschiavo** [comment]
> It's fixed in the next release.

Hi, my friend has this same problem. Do you have an ETA for the stable v0.11? If he installs the new version will he be able to play with me on v0.10.x?
It seems you didn't provide builds for Linux, do I have to compile it myself? How can I do that?

Thank you.

**Comment by Chicken-Bones** [comment]
If you and your friend both obtain the beta release from the discord channel you should be able to test it out for us.

Would be greatly appreciated

**Comment by jtnix** [comment]
Just downloaded and tested v0.11 from discord beta channel, it's working great on MacOS 10.11 and 10.12 (latest versions) !  
Thanks for the update.

**Comment by Jofairden** [comment]
Good to hear, thanks for the update.

**Comment by Thomas-AC-Dev** [comment]
Still crashes on Linux Mint 19.1. Exact same problems listed in main topic. I am using TModLoader 0.10.5.
